packages/solid-query/src/createBaseQuery.ts

```diff
@@ -45,40 +50,109 @@ export function createBaseQuery<
     observer.getOptimisticResult(defaultedOptions),
   )
 
-  const [dataResource, { refetch, mutate }] = createResource<TData | undefined>(
+  const createServerSubscriber = (
+    resolve: (
+      data:
+        | QueryObserverResult<TData, TError>
+        | PromiseLike<QueryObserverResult<TData, TError> | undefined>
+        | undefined,
+    ) => void,
+  ) => {
+    return observer.subscribe((result) => {
+      notifyManager.batchCalls(() => {
+        const unwrappedResult = { ...unwrap(result) }
+
+        setState(unwrappedResult)
+        resolve(unwrappedResult)
+      })()
+    })
+  }
+
+  const createClientSubscriber = (
+    mutate: () => Setter<QueryObserverResult<TData, TError> | undefined>,
+    refetch: () => void,
+  ) => {
+    return observer.subscribe((result) => {
+      notifyManager.batchCalls(() => {
+        const unwrappedResult = { ...unwrap(result) }
+
+        batch(() => {
+          setState(unwrappedResult)
+          mutate()(() => unwrappedResult)
+          refetch()
+        })
+      })()
+    })
+  }
+
+  /**
+   * Unsubscribe is set lazily, so that we can subscribe after hydration when needed.
+   */
+  let unsubscribe: (() => void) | null = null
+
+  const [queryResource, { refetch, mutate }] = createResource<
+    QueryObserverResult<TData, TError> | undefined
+  >(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          unsubscribe = createServerSubscriber(resolve)
+        } else {
+          if (!unsubscribe) {
+            unsubscribe = createClientSubscriber(
+              () => mutate,
+              () => refetch(),
+            )
+          }
+
+          if (!state.isInitialLoading) {
+            resolve(state)
           }
-          resolve(unwrap(state.data))
```

Comment:
TBH, wasn't sure when/where/why to `unwrap` versus not `unwrap`... any rules of thumb?